### PR TITLE
feat(blob): add `clientPayload` and `tokenPayload`

### PR DESCRIPTION
**Pull request by vvo**
This PR allows `clientPayload` during `upload,` so you can attach
metadata to the file uploaded to Vercel Blob. This metadata will be sent as
`tokenPayload` by default in uploading completed callbacks. And you can also enrich
it/configure it/verify it.

Usage:
```ts
// client.ts
import { upload } from '@vercel/blob/client';

upload(
  'file.txt',
  {
    access: 'public',
    handleUploadUrl: '/api/upload',
    clientPayload: JSON.stringify({ postId: 123 })
  }
);
```

```ts
const jsonResponse = await handleUpload({
  body,
  request,
  onBeforeGenerateToken: async (pathname, clientPayload) => {

    return {
       tokenPayload: clientPayload // this is the default, ideally you would verify the clientPayload (format, authorization) before generating the token, so no files are uploaded without first verifying the client claims
    };
  },
  onUploadCompleted: async ({ blob, tokenPayload }) => {
    console.log('blob upload completed', blob, tokenPayload);
  }
});
```

An example use-case of this feature would be to allow editing the images of a
blog post. To do so you would use `clientPayload: JSON.stringify({ postId: 345
})` in the browser. Then, either in onBeforeGenerateToken or in onUploadCompleted
you would verify that the client making the call is allowed to modify this
postId (authorization) and update your database accordingly.

BREAKING CHANGES:
- `metadata` in `onBeforeGenerateToken` and `onUploadCompleted` has been renamed to `tokenPayload`

TODO before release:
- [ ] modify edge-functions to accept tokenPayload AND metadata
- [ ] test on staging for real
- [ ] update front PR documentation

### CODE REVIEW

File: packages/blob/src/client.ts [hunk 1]
```diff
@@ -232,8 +236,12 @@ export async function handleUpload({
   const type = body.type;
   switch (type) {
     case 'blob.generate-client-token': {
-      const { pathname, callbackUrl } = body.payload;
-      const payload = await onBeforeGenerateToken(pathname);
+      const { pathname, callbackUrl, clientPayload } = body.payload;
+      const payload = await onBeforeGenerateToken(pathname, clientPayload);
+      // TODO @Fabio: the next line `?? null` previously, which always made sure that tokenPayload was `null`
```
Comment (vvo) [inline packages/blob/src/client.ts]: cc @correttojs the previous code was : `metadata: payload.metadata ?? null` which ensured the JSON was always `null`, while other types suggest metadata is always either undefined or a string, any reason why?

Comment (correttojs) [reply packages/blob/src/client.ts]: I remember there was something with the metadata serialization. not sure if the `undefined` would cause any issue while serializing/deserializing the client token.
We can add a generateToken unit test passing an empty payload eventually

**Comment by correttojs** [comment]
I just noticed we need to update `edge-functions` to send the `tokenPayload` instead of the `metadata` to the webhook (not the webhook itself luckily)

File: packages/blob/src/client.ts [hunk 2]
```diff
@@ -232,8 +236,12 @@ export async function handleUpload({
...
+      // Why so?
+      const tokenPayload = payload.tokenPayload ?? clientPayload;
```
Comment (correttojs) [inline packages/blob/src/client.ts:243]: that's a bit confusing indeed...we can generate a payload from the server (through `onBeforeGenerateToken`, think the case we need to send the session Id) or from the client (directly through the `clientPayload` for any unsecure data). Which of the two values should have precedence? should we replace this logic by a function, which potentially alows to combine server and client payload?

Comment (correttojs) [reply packages/blob/src/client.ts:243]: oh I see now, we are passing the `clientPayload` to the `onBeforeGenerateToken`. That should be good then (not sure what's the reason of your comment then).

May not be clear for the user that  `onBeforeGenerateToken` can return a tokenPayload...maybe an optional async `onGeneratePayload(clientToken)` would be more clear?

**Review by correttojs — APPROVED** [review]

File: packages/blob/src/client.ts (hunk 1 above)
Comment (vvo) [reply packages/blob/src/client.ts]: Added a unit test, fine

File: packages/blob/src/client.ts (hunk 2 above)
Comment (vvo) [reply packages/blob/src/client.ts:243]: Discussed in call: we keep it this way. The naming internally is not obvious between the various uses of the word "payload" but that's on us.
